### PR TITLE
Avoid dart2js interceptors to improve performance

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Ultimate Problem
In previous profiling, I've found that the lookup of dart2js interceptors involved with interacting with React JS objects takes up a significant amount of rendering time.

See https://github.com/dart-lang/sdk/issues/26247.

## Solution
* Avoid the lookup of dart2js interceptors where possible:
    * Mainly, add custom JS lifecycle methods that retrieve react-dart internals from props objects and pass them to Dart code, avoiding the interceptors the Dart code would've used to access those JS props objects.
    * Move variadic children validation to JS.

These changes only affect `react-dart` internals and should not be breaking.

#### Performance results
Full results are available [in this doc](https://docs.google.com/spreadsheets/d/1GB8QZZGlgFNw4gx6AuRA242Ardfei0sDAYRcHv_osLU/edit#gid=811424833
).

###### tl;dr:
 For 20 levels of components, initial rendering was reduced, in terms of React JS time, from 6.72x to 3.73x.

screen shot 2016-05-26 at 12 43 19 pm

Rendering time is __~56%__ of what it was.

## Testing
* Verify that all tests pass.
* Remove `_createReactDartComponentClassConfig` from a React JS file, load a page with that file, and verify that you get a useful warning.

---

@hleumas @trentgrover-wf @maxpeterson-wf @evanweible-wf @jacehensley-wf

### CODE REVIEW

File: lib/react_client.dart [hunk 1]
```diff
@@ -594,8 +593,11 @@ void setClientConfiguration() {
     // corresponding JS functions are not available.
     React.isValidElement(null);
     ReactDom.findDOMNode(null);
+    createReactDartComponentClassConfig(null, null);
   } on NoSuchMethodError catch (_) {
     throw new Exception('react.js and react_dom.js must be loaded.');
+  } catch(_) {
```
Comment (jacehensley-wf) [inline lib/react_client.dart]: Is there not a specific error type that gets thrown?

File: test/js_interop_helpers_test/js_function_test.dart [hunk 2]
```diff
@@ -38,6 +39,18 @@ void main() {
...
       });
     });
+
+    group('_markChildValidated', () {
+      test('is function that does not throw when called', () {
+        expect(() => markChildValidated(new EmptyObject()), returnsNormally);
+      });
+    });
+
+    group('_markChildValidated', () {
```
Comment (jacehensley-wf) [inline test/js_interop_helpers_test/js_function_test.dart]: Typo should be _createReactDartComponentClassConfig

File: lib/react_client.dart (hunk 1 above)
Comment (greglittlefield-wf) [inline lib/react_client.dart]: No, it's just `Exception` 😞

File: test/js_interop_helpers_test/js_function_test.dart (hunk 2 above)
Comment (greglittlefield-wf) [inline test/js_interop_helpers_test/js_function_test.dart]: Addressed

**Comment by jacehensley-wf** [comment]
+1

**Comment by trentgrover-wf** [comment]
+1
@greglittlefield-wf has this been smoke tested with WSD?

**Comment by greglittlefield-wf** [comment]
@trentgrover-wf Yup, WSD tests pass and everything seems to render properly in the docs.

**Pull request merged**